### PR TITLE
Specialize find and count for vector<bool>

### PR DESCRIPTION
**Pull request by miscco**
This builds uppon the recently merged specialization of `fill` for `vector<bool>`.

Note that the specializations require `__cpp_lib_bitops`

### CODE REVIEW

File: stl/inc/vector
Comment (barcharcraz) [inline stl/inc/vector:2946]: This ifdef can be removed now.

File: stl/inc/vector
Comment (barcharcraz) [inline stl/inc/vector:2955]: what's this actually doing? it's unclear if this is a right shift in disguise or what? Why do we need to use -1 here?

File: stl/inc/vector
Comment (barcharcraz) [inline stl/inc/vector:2966]: I don't love this ?: but I get that it's for a constant so no change requested. a comment may be good though.

File: stl/inc/vector
Comment (barcharcraz) [inline stl/inc/vector:2967]: we might be able to include <bit> directly in vector, but I suppose if you only need _Countr_zero then using this (from limits) is fine.

File: stl/inc/vector
Comment (barcharcraz) [inline stl/inc/vector:2990]: this can probably be removed

File: stl/inc/vector
Comment (barcharcraz) [inline stl/inc/vector:3004]: ditto removed

File: stl/inc/vector
Comment (barcharcraz) [inline stl/inc/vector:3019]: this needs to be removed, builtin_popcount was removed from the compiler. Include <bit> and use std::popcount instead.

File: stl/inc/vector
Comment (barcharcraz) [inline stl/inc/vector:3023]: ditto don't use __builtin_popcount

File: stl/inc/vector
Comment (barcharcraz) [inline stl/inc/vector:3028]: ditto __builtin_popcount

File: stl/inc/vector
Comment (barcharcraz) [inline stl/inc/vector:3034]: ditto __builtin_popcount

File: stl/inc/vector
Comment (barcharcraz) [inline stl/inc/vector:3042]: can probably be removed

File: tests/std/tests/GH_000625_vector_bool_optimization/test.cpp
Comment (barcharcraz) [inline tests/std/tests/GH_000625_vector_bool_optimization/test.cpp:152]: does clang-format really screw this one up?

File: tests/std/tests/GH_000625_vector_bool_optimization/test.cpp
Comment (barcharcraz) [inline tests/std/tests/GH_000625_vector_bool_optimization/test.cpp:153]: is this random?

**Review by barcharcraz — CHANGES_REQUESTED** [review]
reviewed use of bit, and some on functionality. I need to become more familiar with how vector<bool> works before I can give more functional feedback, but this should outline next steps.

Sorry it took me so long to get to this.

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:2946]: Awesome

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:2955]: This is creating a mask, that zeros out the first `_First._Myoff` bits.

`static_cast<_Vbase>(-1) << 2 == [11111111111111111111111111111100]`

This is necessary as we cannot count the previous bits

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:2990]: Done

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:3004]: Ditto

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:3019]: Urgh then I need to pull this into <limits> as we cannot use `<bit>` in `<vector>` without C++20

Comment (miscco) [reply stl/inc/vector:3019]: Looking at it a bit i would guess this effectively blocks this until we can drop the workarounds in bit

`// TRANSITION, VS 2019 16.8 Preview 1, intrin0.h will declare __lzcnt* and __popcnt*`

Otherwise this would need a lot of unnecessary complexity moving a lot of machinery to `<limit>`

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:3042]: Done

File: tests/std/tests/GH_000625_vector_bool_optimization/test.cpp
Comment (miscco) [reply tests/std/tests/GH_000625_vector_bool_optimization/test.cpp:153]: No the first 8 bits are always repeated. There is sufficient irregularity to make it work quite well. Do you have a good idea for some randomly generated sequences?

File: tests/std/tests/GH_000625_vector_bool_optimization/test.cpp
Comment (miscco) [reply tests/std/tests/GH_000625_vector_bool_optimization/test.cpp:152]: Yes, I wanted it to have the byte sequen ces aligned so that it is easier to reason about what the result should be

File: tests/std/tests/GH_000625_vector_bool_optimization/test.cpp
Comment (barcharcraz) [reply tests/std/tests/GH_000625_vector_bool_optimization/test.cpp:153]: No, I'd just like a comment either saying it's random or saying how it's not.

File: stl/inc/vector
Comment (barcharcraz) [reply stl/inc/vector:2954]: oh I see, perhaps numeric_limits<_Vbase>::max() << 2 would be clearer, or a comment to that effect. I don't know why I thought it was a left shift, it obviously isn't, but I guess on quick reading I swapped the left and right side of the <<.

File: stl/inc/vector
Comment (barcharcraz) [reply stl/inc/vector:3011]: we could also pull out the implementation as _Ugly helpers but keep them in bit, above the CXX20 defines. that would probably be better here, but I'll have to poke another maintainer for our standard practice here. If I recall the reasoning for moving what we did to <limits> was because limits is included everywhere and we really didn't want to make it bigger. Vector is already big.

Comment (barcharcraz) [reply stl/inc/vector:3011]: also __popcnt is already in intrin0, but won't help you because you still need to make sure that the architecture actually supports it. __builtin_popcount was to do the fallback for you (essentially being an implementation of std::popcount) as it does in clang and (I assume) gcc, but we don't have any other builtins that work that way and it turns out it was a huge pain to add the first one. So we didn't.

**Comment by miscco** [comment]
@barcharcraz can you give an update on the status of the builtins?

File: tests/std/tests/GH_000625_vector_bool_optimization/test.cpp
Comment (barcharcraz) [reply tests/std/tests/GH_000625_vector_bool_optimization/test.cpp:152]: yeah that's fair

**Comment by barcharcraz** [comment]
> @barcharcraz can you give an update on the status of the builtins?

we have no intention of adding the builtins, and they have been removed. We found the current way we implement these functions to be better given how our compiler is put together.

You'll have to move popcount to some header that vector can include.

File: tests/std/tests/GH_000625_vector_bool_optimization/test.cpp
Comment (miscco) [reply tests/std/tests/GH_000625_vector_bool_optimization/test.cpp:153]: Done

**Comment by miscco** [comment]
Rebased and squashed down now using std::popcount

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector]: Ditto _STD popcount

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector]: Ditto _STD popcount

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector]: Ditto _STD popcount

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector]: Using  _STD popcount

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:2963]: Hm I would keep it as is, because it is both existing practise and all the bit fiddling is nontrivial so to be consitent I would need to massivly comment everything

**Comment by miscco** [comment]
@barcharcraz I believe this is ready for another round of review

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:2974]: I am wondering whether a lambda would help here:
```cpp
auto _Maybe_flip = [](const bool _No_flip, const _VBITS _Mask) -> _VBITS { return _No_flip ? _Mask : ~_Mask; };
```
it should not affect throughput or performance. WDYT?

File: stl/inc/vector
Comment (statementreply) [inline stl/inc/vector:2960]:
```suggestion
    const _Vbase* _VbFirst      = _First._Myptr;
```

File: stl/inc/vector
Comment (statementreply) [inline stl/inc/vector:3010]:
```suggestion
    const _Vbase* _VbFirst      = _First._Myptr;
```

**Comment by StephanTLavavej** [comment]
I've pushed a merge with `master`. There was a merge conflict in `<bit>` caused by clang-format 11 adjusting whitespace in code you removed; I accepted your change.

**Comment by miscco** [comment]
Thanks a lot, I was 99% sure that i already merge master.
Seems like not pushing is a bad idea

**Comment by miscco** [comment]
Urgh, Seems I has an unclean state and the use `cbegin `for `fill `commit is still there 🤦

File: stl/inc/vector
Comment (barcharcraz) [reply stl/inc/vector:3197]: I prefer the ternary to the lambda

File: stl/inc/vector
Comment (barcharcraz) [reply stl/inc/vector:3186]: yeah that's fine.

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:3197]: I am super glad you do

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:3198]: Is this still an issue?

**Comment by miscco** [comment]
@barcharcraz are there any further blockers here?

**Review by barcharcraz — APPROVED** [review]
Approved with suggestions (it still needs final review). Also: before / after benchmarks would be awesome.

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:3177]: Renamed to `_VbIt` in all functions

File: stl/inc/vector
Comment (miscco) [reply stl/inc/vector:3178]: I would keep it as we have it in all others

**Comment by miscco** [comment]
So I ran the itsy.bitsy benchmarks from the wonderfull @ThePhD

The relevant results are for current version:
```
Running C:\ws\itsy_bitsy\build\x64\bin\Release\itsy.bitsy.benchmarks.exe
Run on (8 X 1992 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 8192 KiB (x1)
--------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations
--------------------------------------------------------------------------
count_vector_bool                    95765 ns        96257 ns         7467
count_itsy_bitsy                       842 ns          854 ns       896000
find_vector_bool                     84993 ns        85449 ns         8960
find_itsy_bitsy                      49712 ns        50000 ns        10000
```

With this PR:

The relevant results are for current version:
```
Running C:\ws\itsy_bitsy\build\x64\bin\Release\itsy.bitsy.benchmarks.exe
Run on (8 X 1992 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 8192 KiB (x1)
--------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations
--------------------------------------------------------------------------
count_vector_bool                     4666 ns         4353 ns       186667
count_itsy_bitsy                      1607 ns         1562 ns       640000
find_vector_bool                      3205 ns         3209 ns       224000
find_itsy_bitsy                      50459 ns        51562 ns        10000
```

So we are still losing out against itsy.bitsy, which is not unexpected. But the performance increase is
```
count:    95'000 ns -> 5'000ns -> 19x improvement
find:     85'000ns  -> 3'200ns -> 26x improvement
```

I believe this is no too shabby

**Comment by miscco** [comment]
@barcharcraz I added the optimizations to `ranges::count` and `ranges::find`

File: tests/std/tests/GH_000625_vector_bool_optimization/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/GH_000625_vector_bool_optimization/test.cpp:175]: We conventionally avoid C casts:
```suggestion
    const auto expected = div(static_cast<int>(length), 8);
```

Comment (StephanTLavavej) [inline tests/std/tests/GH_000625_vector_bool_optimization/test.cpp:175]: Should include `<cstdlib>` for `div`.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/limits
Comment (miscco) [reply stl/inc/limits:1163]: Oh that is a great observation. `popcount` is always constexpr and I was simply moving the code around.

**Review by StephanTLavavej — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Internal testing revealed that `/clr:pure` rejects this intrinsic (as it does for almost all intrinsics). I belatedly realized that CUDA and the Intel C++ Compiler may also be incompatible. I've extracted these conditions into a macro and verified that `/clr:pure` is happy now.

`_Countr_zero`'s usage of `tzcnt` is unaffected because it behaves differently - it uses the `constexpr` fallback for C++17 mode, and only for C++20 mode (when `is_constant_evaluated()` is available) does it consider activating the runtime optimization. This inherently excludes `/clr:pure` and also excludes CUDA for the time being (and possibly the Intel C++ Compiler).

**Comment by StephanTLavavej** [comment]
Thanks for these order-of-magnitude speedups! :rocket: :rocket: :rocket:

**Pull request merged**
